README.md
# pyren
pyren is a program for getting diagnostic data from Renault cars by ELM327 adapter. It works in two modes and with two types of databases accordingly.
Compatible with Windows, Linux,  MacOS and Android(only CLIP mode under Android)

## CLIP mode
You may use this mode when you have installed original CLIP software but don't have CLIP hardware

This mode allows you:
* Choose screens languages
* Scan and connect to any ECU of your car
* Read diagnostic data from any ECU
* Read DTC codes from ECUs and get their user friendly descriptions compliant with documentation
* Clear DTC 
* Change configuration (depends on ECU)
* Execute equipment tests (depends on ECU)

## DDT mode
You may use this mode when you have DDT2000 database. It may be run standalone or from CLIP mode. In case of running from CLIP mode it allows to translate some parameters descriptions to choosen language.

This mode allows you:
* Get full control under all ECUs parameters (be very careful in expert mode)
* Save dump of ECU configuration. It may allows you to revert configuration back in case of trouble after configuring
* Compare configurations
* Rollback configuration

## Included tools
* **doc_maker.py** - build diagnostic documentation for your car (require BVMEXTRACTION folder and extracted DocDB_xx.7ze)
* **bus_monitor.py** - catch, parse and shows the content of system frames on your CAN bus.
* **mod_optfile.py** - for exploring CLIP database
* **mod_term.py** - simple terminal
* **mod_ecu.py** - shows list of all parameters, commands and DTCs. Build "PID" CSV file for "Torque PRO" 

# Installation

## Dependencies
* **Python 2.7.xx**
* **ELM327** adapter with FlowControl support (original one or chinese v1.5)
    * **ELM327-USB** - preferred for Windows, Linux and MacOS
    * **ELM327-BT** - not compatible with Android 7.0 and above, use ELM327-WiFi
* **CLIP database** required for CLIP mode, not included (not included, you must own it)
* **DDT2000 database** required for DDT mode, not included (not included, you must own it)

Getting started video
https://www.youtube.com/watch?v=DazsvlnLCoA

## Installation on Windows

Check that you have CLIP database installed 
```
cd c:\CLIP\Data\GenAppli\
dir 
```
This directory should contains at least the next subdirectories
* EcuRenault
* Location
* Vehicles

Install pyren
```
cd c:\CLIP\Data\GenAppli\
git clone git@gitlab.com:py_ren/pyren.git
```
or just download it

https://gitlab.com/py_ren/pyren/-/archive/master/pyren-master.zip

and extract next to directories EcuRenault, Location, Vehicles

You have to get the next directory tree
```
<any work directory>
|- BVMEXTRACTION            #(need for doc_maker)
|- DocDB_xx                 #(need for doc_maker where xx=language) 
|- EcuRenault               #(for CLIP mode)
|- Location                 #(for CLIP mode)
|- Vehicles                 #(for CLIP mode)
|- ecus                     #(for DDT mode)
|- graphics                 #(for DDT mode optional)
|- pyren                    #(pyren)
|   |- pyren.py
...    ...
|   |- <other modules>
|- _pyren_launcher.py       #(universal launcher)
```

## Installation on MacOS and Linux

For running under MacOS or Linux you need just to copy the same directory tree as in Windows installation above

## Installation on Android

1. Install SL4A and Py4A. (https://github.com/kuri65536/python-for-android/blob/master/README.md)
2. Copy DB folders and the pyren into sl4a/scripts. You have to get the next direcotry tree

```
sl4a/scripts
|- EcuRenault           
|- Location             
|- Vehicles             
|- pyren              
|   |- pyren.py
...    ...
|   |- <other modules>
|- _pyren_launcher.py   
```


# User guide

**BE VERY CAREFUL. ONLY YOU RESPONSIBLE FOR ALL YOU ARE DOING WITH YOUR CAR**

## Running from universal launcher

Universal launcher **_pyren_launcher.py** is compatible with any OS. Run it under Windows, MacOS, Linux or Android

## Running from command line

Firs run pyren without options

```
$cd c:\CLIP\Data\GenAppli\pyren
$python.exe ./pyren.py 
usage: pyren.py [-h] [-v] [-p PORT] [-s SPEED] [-r RATE] [-L LANG] [-m CAR]
                [-vv] [-e ECUID] [--si] [--cfc] [--n1c] [--csv] [--csv_only]
                [--csv_human] [--usr_key USR_KEY] [--log LOGFILE] [--scan]
                [--demo] [--dump] [--dev DEV] [--exp] [--can2]

pyRen - python program for diagnostic Renault cars

optional arguments:
  -h, --help         show this help message and exit
  -v, --version      show program's version number and exit
  -p PORT            ELM327 com port name
  -s SPEED           com port speed configured on ELM
                     {38400[default],57600,115200,230400,500000} DEPRECATED
  -r RATE            com port rate during diagnostic session
                     {38400[default],57600,115200,230400,500000}
  -L LANG            language option {RU[default],GB,FR,IT,...}
  -m CAR             number of car model
  -vv, --verbose     show parameter explanations
  -e ECUID           index of ECU, or comma separeted list for DEMO MODE
  --si               try SlowInit first
  --cfc              turn off automatic FC and do it by script
  --n1c              turn off L1 cache
  --csv              save data in csv format
  --csv_only         data doesnt show on screen for speed up
  --csv_human        data saves to csv in readable format
  --usr_key USR_KEY  add user events to log
  --log LOGFILE      log file name
  --scan             scan ECUs even if savedEcus.p file exists
  --demo             for debuging purpose. Work without car and ELM
  --dump             dump responces from all 21xx and 22xxxx requests
  --dev DEV          swith to Development Session for commands from DevList,
                     you should define alternative command for opening the
                     session, like a 1086
  --exp              swith to Expert mode (allow to use buttons in DDT)
  --can2             CAN network connected to pin 13 and pin 12

Available COM ports:
COM3:              
	desc: n/a 
	hwid: n/a

$

```

it shows the list of available options and COM ports

Run pyren with a mandatory option {-p PORT} 

```
$cd c:\CLIP\Data\GenAppli\pyren
$python.exe pyren.py -p COM3
``` 

For the first start you will be prompted to choose a model of car

```
$python.exe pyren.py -p COM3
Opening ELM
Loading ECUs list

1  - B90 SANDERO
2  - J77 MODUS
3  - XDXX MASTER ph2
4  - X95 MEGANEIII/SCENICIII
5  - X85 CLIO III
6  - X24 MASTER PROP.
7  - X81X ESPACE IV
8  - X65 CLIO V6 ph2
9  - X74X LAGUNA II
10 - X70 MAST.PRO.ph2
11 - X44 TWINGO II
12 - X81 P2 ESPACEIV ph2/3/4
13 - XFD TALISMAN/SM6
14 - XFA SCENIC IV
15 - XFF MEGANE IV SEDAN
16 - X79 DUSTER ph2
17 - XHA CAPTUR/KAPTUR BR/IN/RU
18 - XZJ KOLEOS II CN/QM6 CN
19 - XZG KOLEOS II/QM6
20 - XFE KADJAR
N  - <next page>
Choose model :
```

Type the number according your model and press {Enter}. For example, if your car is SANDERO then you have to type '1' and press <Enter>

```
$python.exe pyren.py -p COM3
Opening ELM
Loading ECUs list

1  - B90 SANDERO
2  - J77 MODUS
3  - XDXX MASTER ph2
4  - X95 MEGANEIII/SCENICIII
5  - X85 CLIO III
6  - X24 MASTER PROP.
7  - X81X ESPACE IV
8  - X65 CLIO V6 ph2
9  - X74X LAGUNA II
10 - X70 MAST.PRO.ph2
11 - X44 TWINGO II
12 - X81 P2 ESPACEIV ph2/3/4
13 - XFD TALISMAN/SM6
14 - XFA SCENIC IV
15 - XFF MEGANE IV SEDAN
16 - X79 DUSTER ph2
17 - XHA CAPTUR/KAPTUR BR/IN/RU
18 - XZJ KOLEOS II CN/QM6 CN
19 - XZG KOLEOS II/QM6
20 - XFE KADJAR
N  - <next page>
Choose model : 1 <Enter>
```
Type 'N' and {Enter} for open next page of list

Then pyren starts ECU scanning

```
Choose model :1
Loading data for : B90 SANDERO ../Vehicles/TCOM_139.xml   - 62 ecus loaded
Scanning:53/62 Detected: 4      
```

List of all ecus will be saved in file **savedEcus.p** for not to scan them every time. For connecting to a new car or rescan your car, use option --scan or just delete file savedEcus.p

Also I suggest you to use option **--log** and **--dump** every time you start the pyren.

After first connection to ECU with **--dump** option you will be able to run the pyren offline in demo mode

 ```
$cd c:\CLIP\Data\GenAppli\pyren
$python.exe pyren.py -pp --demo
``` 

(As so **-p** is a mandatory option you may define any value for demo)




pyren/CHANGE_LOG.txt
История версий


v 0.8 (beta)

- Первая версия от PyRen. Основана на версии скрипта pyren light v 0.7 (alpha) от Shr-lnm “club-renault.ru”
- Реализована работа с базами клип

v 0.9 (beta)

Исправлено:
- По результатам тестирования на Koleos исправлена пара ошибок иногда проявляющихся в меню DE
Добавлено:
- Упрощена инсталляция под Windows. Теперь не нужно ставить дополнительные модули Python. Скрипт сам их поставит при первом запуске. (Компьютер должен быть в сети)
- Разблокированы простые команды не требующие параметров (используйте с осторожностью)
- После сканирования, меню выбора ЭБУ показывает примерное количество ошибок. (Примерное потому, что некоторые не существенные ошибки далее могут игнорироваться)

v 0.9.1 (beta)

Исправлено:
- на Windows машинах, в меню некоторых блоков, вываливался с ошибкой отсутствия символа в кодовой странице cp866
- не правильно показывал состояния и параметры длина которых больше 1-го и меньше 8-ми бит (поля с длиной более байта но не кратно 8-ми бит (если такие есть) по прежнему обрабатываться правильно не будут)
- на некоторых блоках вываливался в меню ошибок
- показывал не все ошибки в блоках STD_B
- некорректно отображалось буквенное обозначение параметров и состояний
Добавлено:
- В меню выбора ЭБУ добавлен пункт повторного сканирования ошибок по всем обнаруженным блокам.

v 0.9.2 (beta)

Исправлено:
- на windows машинах вываливался при нажатии не буквенных клавиш.
- вываливался при загрузке блоков с некорректно описанными переводами в языковой базе. Теперь,  в таких случаях, берется поле defaulttext. Можно самостоятельно исправлять поле defaulttext если сделать не валидным значение поля codetext.
- исправлена ошибка повторного сканирования
Добавлено:
- модуль mod_optfile.py можно запускать самостоятельно для исследования содержимого файлов bqm и SGxxxxxxx.XML

v 0.9.3 (beta)

Исправлено:
- исправления в алгоритме сканирования и работы с к-линией (исправление от Shr-lnm)
- выход из любого пункта меню осуществляется кнопкой Q
- если запустить скрипт с ключом -L и указать не существующий язык, например FR, то языковая база не загрузится и все сообщения будут на языке оригинала (максимально быстрая загрузка)
- несколько мелких багов приводивших к вываливанию скрипта
Добавлено:
- поддержка wifi адаптеров. вместо названия com порта "-p com1" нужно указать адрес и порт wifi адаптера, например "-p 192.168.0.10:35000". Некоторые модели wifi адаптеров работают очень медленно - см. соответствующие форумы по настройкам адаптеров.
- поддержка серии сервисов в одной команде (когда одна команда меню вызывает последовательность простык команд ЭБУ)

v 0.9.4 (beta)

Исправлено:
- алгоритм изменения интервалов между сериями команд в соответсвии с указаниями из БД
- загрузка некоторых блоков прерывалась на стадии Loading defaults
Добавлено:
- выполенение серии команд без парамаметров

v 0.9.5 (beta)

Исправлено:
- при работе по CAN на ELM отключается автоворматирование “AT CAF0” это позволяет подавать команды длинее 7 байт (исправление от Shr-lnm)
- отключен ограничитель частоты опроса ЭБУ (busLoad) (исправление от Shr-lnm)
- алгоритм работы команд требующих повторения.
Добавлено:
- опция -r <baudrate> позволяет изменить скорость работы COM порта во время работы скрипта. (добавил Shr-lnm)
- Основное меню ЭБУ показывает описание пункта дополнительно к двухбуквенному (предложил gruzdev_f)
- Работа с командами, параметры которых выбираются из списка. (Например: смена языка в панели приборов, активирование круиз контроля и пр.)
- Ключ --csv включает запись параметров и состояний в CSV файл для дальнейшего анализа, например в exel. 
- mod_ecu.py стал запускаемым. С его помощью можно посмотреть полный список параметров и команд ЭБУ

v 0.9.6 (beta)

Исправлено:
- несколько косметических ошибок.
Добавлено:
- быстрая загрузка (со второго раза).
- отключен автоматический FlowControl “AT CFC0” (добавил Shr-lnm)

v 0.9.7 (beta)

Исправлено:
- последовательности инициализации ELM для CAN и K-линии
- по умолчанию автоматический FlowControl не отключается
- отображение хода выполнения команд состоящих из нескольких сервисов
- перед считыванием ошибок таймер увеличивается до 1 секунды
- если скорость порта меньше 50kbps то команда 1902хх заменяется на 1902AF
- keepalive включается на CAN и ISO
Добавлено:
- поддержка нескольких сервисов в одной мнемонике, данные берутся из последнего сервиса

v 0.9.8 (beta)

Исправлено:
- обработка параметров с обратным порядком байт
- последовательность инициализации k-line (Shr-lnm)
Добавлено:
- поддержка запуска под Android
- генерация CSV для torque (Shr-lnm)

v 0.9.9 (beta)

Исправлено:
- более корректно работает mod_optfile.py (спасибо Valentin8080)
- контроль длины вводимого параметра HEX
- изменена нумерация моделей автомобилей
- при выборе модели не показываются старые, неподдерживаемые модели. 
- длинные меню адаптированы под маленький экран андроид
Добавлено:
- поддержка FAST/SLOW INIT на k-line (kline-mode2 от Shr-lnm)
- значительно улучшены возможности commander.py (см. пример кода)
- commander.py переименован в cmdr_example.py и изменен launcher
- в меню просмотра ошибок блоков STD_B выдаются дополнительные параметры (спасибо Ivaness)
- добавлен скрипт cmdr_odometr.py для проверки пробега в разных блоках

v 0.9.9(3) (beta)

Исправлено:
- алгоритм чтения ответов ELM (mod_elm_odometr_v2 от Shr-lnm)
- время отклика в логах
- ошибка логирования под Windows (спасибо Vitna)
- ошибка чтения некоторых параметров под Android (спасибо Ivaness)
Добавлено:
- принудительное включение SlowInit. Oпция --si
- отключение автоматического FlowControl выполняемого ELM. С опцией --cfc скрипт сам будет делать FC
- отключение ускорения работы по CAN шине за счет указания количества фреймов в ответе. Опция --n1c отключает это ускорение
- учет параметра brp (автоматический выбор CAN 250k/500k)

v 0.9.9(4) (beta)

Исправлено:
- работа опции --cfc

v 0.9.9(5) (beta)

Исправлено:
- выполнение некоторых команд.

pyren/pyren.py
#!/usr/bin/env python

import sys, os
import mod_globals

mod_globals.os = os.name

os.chdir(os.path.dirname(os.path.realpath(sys.argv[0])))

import pickle

if mod_globals.os == 'nt':
  import pip
  
  try:
    import serial
  except ImportError:
    pip.main(['install','pyserial'])

  try:
    import colorama
  except ImportError:
    pip.main(['install','colorama'])
    try:
      import colorama
    except ImportError:
      print "\n\n\n\t\t\tGive me access to the Internet for download modules\n\n\n"
      sys.exit()
  colorama.init()
else:
  # let's try android
  try:
    import androidhelper as android
    mod_globals.os = 'android'
  except:
    try:
      import android
      mod_globals.os = 'android'
    except:
      pass
  
if mod_globals.os != 'android':    
  try:
    import serial
    from serial.tools  import list_ports
    #import ply
  except ImportError:
    print "\n\n\n\tPleas install additional modules"
    print "\t\t>sudo easy_install pyserial"
    #print "\t\t>sudo easy_install ply"
    sys.exit()
    
import mod_utils
import mod_ddt_utils

from mod_elm       import ELM
from mod_scan_ecus import ScanEcus
from mod_ecu       import ECU
from mod_optfile   import *
from mod_utils     import *


def optParser():
  '''Parsing of command line parameters. User should define at least com port name'''
  
  import argparse

  parser = argparse.ArgumentParser(
    #usage = "%prog -p <port> [options]",
    version="pyRen Version 0.9.p",
    description = "pyRen - python program for diagnostic Renault cars"
  )
  
  parser.add_argument('-p',
      help="ELM327 com port name",
      dest="port",
      default="")

  parser.add_argument("-s",
      help="com port speed configured on ELM {38400[default],57600,115200,230400,500000} DEPRECATED",
      dest="speed",
      default="38400")

  parser.add_argument("-r",
      help="com port rate during diagnostic session {38400[default],57600,115200,230400,500000}",
      dest="rate",
      default="38400",)

  parser.add_argument("-L",
      help="language option {RU[default],GB,FR,IT,...}",
      dest="lang",
      default="RU")

  parser.add_argument("-m",
      help="number of car model",
      dest="car",
      default=0,
      type=int)

  parser.add_argument("-vv", "--verbose",
      help="show parameter explanations",
      dest="verbose",
      default=False,
      action="store_true")

  parser.add_argument("-e",
      help="index of ECU, or comma separeted list for DEMO MODE",
      dest="ecuid",
      default="")

  parser.add_argument("--si",
      help="try SlowInit first",
      dest="si",
      default=False,
      action="store_true")

  parser.add_argument("--cfc",
      help="turn off automatic FC and do it by script",
      dest="cfc",
      default=False,
      action="store_true")

  parser.add_argument("--n1c",
      help="turn off L1 cache",
      dest="n1c",
      default=False,
      action="store_true")

  parser.add_argument("--csv",
      help="save data in csv format",
      dest="csv",
      default=False,
      action="store_true")

  parser.add_argument("--csv_only",
      help="data doesnt show on screen for speed up",
      dest="csv_only",
      default=False,
      action="store_true")

  parser.add_argument("--csv_human",
      help="data saves to csv in readable format",
      dest="csv_human",
      default=False,
      action="store_true")

  parser.add_argument('--usr_key',
      help="add user events to log",
      dest="usr_key",
      default="")

  parser.add_argument("--log",
      help="log file name",
      dest="logfile",
      default="")

  parser.add_argument("--scan",
      help="scan ECUs even if savedEcus.p file exists",
      dest="scan",
      default=False,
      action="store_true")

  parser.add_argument("--demo",
      help="for debuging purpose. Work without car and ELM",
      dest="demo",
      default=False,
      action="store_true")

  parser.add_argument("--dump",
      help="dump responces from all 21xx and 22xxxx requests",
      dest="dump",
      default=False,
      action="store_true")

  parser.add_argument("--dev",
      help="swith to Development Session for commands from DevList, you should define alternative command for opening the session, like a 1086",
      dest="dev",
      default='')

  parser.add_argument("--exp",
      help="swith to Expert mode (allow to use buttons in DDT)",
      dest="exp",
      default=False,
      action="store_true")

  parser.add_argument("--can2",
      help="CAN network connected to pin 13 and pin 12",
      dest="can2",
      default=False,
      action="store_true")

  options = parser.parse_args()
  
  if not options.port and mod_globals.os != 'android':
    parser.print_help()
    iterator = sorted(list(list_ports.comports()))
    print ""
    print "Available COM ports:"
    for port, desc, hwid in iterator:
      print "%-30s \n\tdesc: %s \n\thwid: %s" % (port,desc.decode("windows-1251"),hwid)
    print ""
    exit(2)
  else:
    mod_globals.opt_port      = options.port
    mod_globals.opt_ecuid     = options.ecuid
    mod_globals.opt_speed     = int(options.speed)
    mod_globals.opt_rate      = int(options.rate)
    mod_globals.opt_lang      = options.lang
    mod_globals.opt_car       = options.car
    mod_globals.opt_log       = options.logfile
    mod_globals.opt_demo      = options.demo
    mod_globals.opt_scan      = options.scan
    mod_globals.opt_csv       = options.csv
    mod_globals.opt_csv_only  = options.csv_only
    if mod_globals.opt_csv_only  : mod_globals.opt_csv = True
    mod_globals.opt_csv_human = options.csv_human
    if mod_globals.opt_csv_human : mod_globals.opt_csv = True
    mod_globals.opt_usrkey    = options.usr_key
    mod_globals.opt_verbose   = options.verbose
    mod_globals.opt_si        = options.si
    mod_globals.opt_cfc0      = options.cfc
    mod_globals.opt_n1c       = options.n1c
    mod_globals.opt_exp       = options.exp
    mod_globals.opt_dump      = options.dump
    mod_globals.opt_can2      = options.can2
    if options.dev=='' or len(options.dev)!=4 or options.dev[0:2]!='10':
      mod_globals.opt_dev       = False
      mod_globals.opt_devses    = '1086'   
    else:
      print "Development MODE"
      mod_globals.opt_dev       = True
      mod_globals.opt_devses    = options.dev   
    
def main():
  '''Main function'''

  optParser()

  mod_utils.chkDirTree()

  print 'Opening ELM'
  elm = ELM( mod_globals.opt_port, mod_globals.opt_speed, mod_globals.opt_log )

  #change serial port baud rate 
  if mod_globals.opt_speed<mod_globals.opt_rate and not mod_globals.opt_demo:
    elm.port.soft_boudrate( mod_globals.opt_rate )
     
  print 'Loading ECUs list'
  se  = ScanEcus(elm)                    #Prepare list of all ecus
 
  SEFname = "savedEcus.p" 
  if mod_globals.opt_can2:
    SEFname = "savedEcus2.p" 

  if mod_globals.opt_demo and len(mod_globals.opt_ecuid)>0:
    # demo mode with predefined ecu list
    se.read_Uces_file( all = True )    
    se.detectedEcus = []
    for i in mod_globals.opt_ecuid.split(','):
      if  i in se.allecus.keys():
        se.allecus[i]['ecuname']=i
        se.allecus[i]['idf']=se.allecus[i]['ModelId'][2:4]
        if se.allecus[i]['idf'][0]=='0': 
          se.allecus[i]['idf'] = se.allecus[i]['idf'][1]
        se.allecus[i]['pin'] = 'can' 
        se.detectedEcus.append( se.allecus[i] )    
  else:
    if not os.path.isfile(SEFname) or mod_globals.opt_scan: 
      # choosing model 
      se.chooseModel( mod_globals.opt_car )  #choose model of car for doing full scan
    
    # Do this check every time
    se.scanAllEcus()                       #First scan of all ecus
 
    
  print "Loading language "
  sys.stdout.flush()
                                         #loading language data
  lang = optfile("../Location/DiagOnCan_"+mod_globals.opt_lang+".bqm",True)
  mod_globals.language_dict = lang.dict
  print "Done"
  
  #check cache version
  if not os.path.isfile("./cache/version09k.txt"):
    #if the cache has old version then we should clear it
    for root, dirs, files in os.walk("./cache"):
      for sfile in files:
        if sfile.startswith("ver") or sfile.startswith("FG") or sfile.startswith("ddt"):
          full_path = os.path.join("./cache", sfile)
          os.remove(full_path)
    #create new version file
    verfile = open( "./cache/version09k.txt", "wb" )
    verfile.write("Do not remove me if you have v0.9.k or above.\n")
    verfile.close()

  mod_ddt_utils.searchddtroot()

  #check if DDT present
  if os.path.exists(mod_globals.ddtroot+'/ecus') and mod_globals.os != 'android':
    mod_globals.opt_ddt = True  

  while( 1 ):
    clearScreen()
    choosen_ecu = se.chooseECU( mod_globals.opt_ecuid )   # choose ECU among detected
    mod_globals.opt_ecuid = ''
    if choosen_ecu==-1:
      continue
    
    ecucashfile = "./cache/"+choosen_ecu['ModelId']+'_'+mod_globals.opt_lang+".p"
    
    if os.path.isfile(ecucashfile):                         #if cache exists
      ecu = pickle.load( open( ecucashfile, "rb" ) )        #load it
    else:                                                   #else
      ecu = ECU(choosen_ecu, lang.dict )                    #loading original data for chosen ECU
      pickle.dump( ecu, open( ecucashfile, "wb" ) )         #and save cache
    
    ecu.initELM( elm )                                      #init ELM for chosen ECU 
    
    if mod_globals.opt_demo:
      print "Loading dump"
      ecu.loadDump()
    elif mod_globals.opt_dump:
      print "Saving dump"
      ecu.saveDump()
      
    ecu.show_screens()                                      #show ECU screens

if __name__ == '__main__':  
  main()
  
